README.md
# turf-deintersect

[![build status](https://secure.travis-ci.org/stepankuzmin/turf-deintersect.png)](http://travis-ci.org/stepankuzmin/turf-deintersect)
[![Greenkeeper badge](https://badges.greenkeeper.io/stepankuzmin/turf-deintersect.svg)](https://greenkeeper.io/)

Delete GeoJSON polygon intersections


### `turf.deintersect(polygons)`

Delete Polygon|polygons intersections by clipping from next polygon
each previous polygon.

### Parameters

| parameter  | type               | description            |
| ---------- | ------------------ | ---------------------- |
| `polygons` | Array\.\<Polygon\> | input polygon features |


### Example

```js
var poly1 = {
  "type": "Feature",
  "properties": {
    "fill": "#0f0"
  },
  "geometry": {
    "type": "Polygon",
    "coordinates": [[
      [-46.738586, -23.596711],
      [-46.738586, -23.458207],
      [-46.560058, -23.458207],
      [-46.560058, -23.596711],
      [-46.738586, -23.596711]
    ]]
  }
};
var poly2 = {
  "type": "Feature",
  "properties": {
    "fill": "#00f"
  },
  "geometry": {
    "type": "Polygon",
    "coordinates": [[
      [-46.650009, -23.631314],
      [-46.650009, -23.5237],
      [-46.509246, -23.5237],
      [-46.509246, -23.631314],
      [-46.650009, -23.631314]
    ]]
  }
};

var deintersected = turf.deintersect([poly1, poly2]);

var polygonsBefore = {
  "type": "FeatureCollection",
  "features": [poly1, poly2]
};

var polygonsAfter = {
  "type": "FeatureCollection",
  "features": deintersected
};

//=polygonsBefore

//=polygonsAfter
```


**Returns** `Array.<Polygon>`, a polygon features without intersections

## Installation

Requires [nodejs](http://nodejs.org/).

```sh
$ npm install turf-deintersect
```

## Tests

```sh
$ npm test
```




CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.4"></a>
## [1.0.4](https://github.com/stepankuzmin/turf-deintersect/compare/v1.0.3...v1.0.4) (2018-09-20)


### Bug Fixes

* filter out empty difference results ([57b22e6](https://github.com/stepankuzmin/turf-deintersect/commit/57b22e6))



<a name="1.0.3"></a>
## [1.0.3](https://github.com/stepankuzmin/turf-deintersect/compare/v1.0.2...v1.0.3) (2018-04-02)


### Bug Fixes

* **package:** update [@turf](https://github.com/turf)/difference to version 5.0.0 ([187f7a4](https://github.com/stepankuzmin/turf-deintersect/commit/187f7a4))
* **package:** update [@turf](https://github.com/turf)/difference to version 5.0.4 ([e372e7c](https://github.com/stepankuzmin/turf-deintersect/commit/e372e7c))
* **package:** update [@turf](https://github.com/turf)/difference to version 5.0.5 ([9de2483](https://github.com/stepankuzmin/turf-deintersect/commit/9de2483))
* **package:** update [@turf](https://github.com/turf)/difference to version 5.1.0 ([07aa759](https://github.com/stepankuzmin/turf-deintersect/commit/07aa759))
* **package:** update [@turf](https://github.com/turf)/difference to version 5.1.1 ([56ba1a1](https://github.com/stepankuzmin/turf-deintersect/commit/56ba1a1))
* **package:** update [@turf](https://github.com/turf)/difference to version 5.1.5 ([a764eff](https://github.com/stepankuzmin/turf-deintersect/commit/a764eff))



<a name="1.0.2"></a>
## [1.0.2](https://github.com/stepankuzmin/turf-deintersect/compare/v1.0.0...v1.0.2) (2017-10-05)


### Bug Fixes

* **package:** update [@turf](https://github.com/turf)/difference to version 4.4.0 ([be2c4cd](https://github.com/stepankuzmin/turf-deintersect/commit/be2c4cd))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.5.1 ([04d8773](https://github.com/stepankuzmin/turf-deintersect/commit/04d8773))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.6.0 ([9b4b081](https://github.com/stepankuzmin/turf-deintersect/commit/9b4b081))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.7.4 ([2301bc8](https://github.com/stepankuzmin/turf-deintersect/commit/2301bc8))



<a name="1.0.1"></a>
## [1.0.1](https://github.com/stepankuzmin/turf-deintersect/compare/v1.0.0...v1.0.1) (2017-05-15)



<a name="1.0.0"></a>
# 1.0.0 (2017-05-15)


### Bug Fixes

* **package:** update [@turf](https://github.com/turf)/difference to version 3.10.0 ([d9bb429](https://github.com/stepankuzmin/turf-deintersect/commit/d9bb429))
* **package:** update [@turf](https://github.com/turf)/difference to version 3.10.3 ([810ba5a](https://github.com/stepankuzmin/turf-deintersect/commit/810ba5a))
* **package:** update [@turf](https://github.com/turf)/difference to version 3.10.4 ([543d447](https://github.com/stepankuzmin/turf-deintersect/commit/543d447))
* **package:** update [@turf](https://github.com/turf)/difference to version 3.13.0 ([e3e1ea9](https://github.com/stepankuzmin/turf-deintersect/commit/e3e1ea9))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.0.0 ([5902005](https://github.com/stepankuzmin/turf-deintersect/commit/5902005))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.1.0 ([30ab085](https://github.com/stepankuzmin/turf-deintersect/commit/30ab085))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.2.0 ([ba3f1e3](https://github.com/stepankuzmin/turf-deintersect/commit/ba3f1e3))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.2.1 ([4e20fcb](https://github.com/stepankuzmin/turf-deintersect/commit/4e20fcb))
* **package:** update [@turf](https://github.com/turf)/difference to version 4.3.0 ([9fc1890](https://github.com/stepankuzmin/turf-deintersect/commit/9fc1890)), closes [#10](https://github.com/stepankuzmin/turf-deintersect/issues/10)


test/test.js
const fs = require("fs");
var test = require("tape");
const difference = require("@turf/difference");
var deintersect = require("../");

var poly1 = {
  type: "Feature",
  properties: {
    fill: "#0f0"
  },
  geometry: {
    type: "Polygon",
    coordinates: [
      [
        [-46.738586, -23.596711],
        [-46.738586, -23.458207],
        [-46.560058, -23.458207],
        [-46.560058, -23.596711],
        [-46.738586, -23.596711]
      ]
    ]
  }
};

var poly2 = {
  type: "Feature",
  properties: {
    fill: "#00f"
  },
  geometry: {
    type: "Polygon",
    coordinates: [
      [
        [-46.650009, -23.631314],
        [-46.650009, -23.5237],
        [-46.509246, -23.5237],
        [-46.509246, -23.631314],
        [-46.650009, -23.631314]
      ]
    ]
  }
};

var result = [
  {
    type: "Feature",
    properties: {
      fill: "#0f0"
    },
    geometry: {
      type: "Polygon",
      coordinates: [
        [
          [-46.738586, -23.596711],
          [-46.738586, -23.458207],
          [-46.560058, -23.458207],
          [-46.560058, -23.596711],
          [-46.738586, -23.596711]
        ]
      ]
    }
  },
  {
    type: "Feature",
    properties: {
      fill: "#00f"
    },
    geometry: {
      type: "Polygon",
      coordinates: [
        [
          [-46.650009, -23.631314],
          [-46.509246, -23.631314],
          [-46.509246, -23.5237],
          [-46.560058, -23.5237],
          [-46.560058, -23.596711],
          [-46.650009, -23.596711],
          [-46.650009, -23.631314]
        ]
      ]
    }
  }
];

test("delete-intersections", function(t) {
  var differenced = deintersect([poly1, poly2]);
  t.deepEqual(differenced, result);
  t.end();
});


index.js
const difference = require("@turf/difference");

/**
 * Delete {@link Polygon|polygons} intersections by clipping from next polygon
 * each previous polygon.
 *
 * @module turf/deintersect
 * @category transformation
 * @param {Array<Polygon>} polygons input polygon features
 * @return {Array<Polygon>} a polygon features without intersections
 * @example
 * var poly1 = {
 *   "type": "Feature",
 *   "properties": {
 *     "fill": "#0f0"
 *   },
 *   "geometry": {
 *     "type": "Polygon",
 *     "coordinates": [[
 *       [-46.738586, -23.596711],
 *       [-46.738586, -23.458207],
 *       [-46.560058, -23.458207],
 *       [-46.560058, -23.596711],
 *       [-46.738586, -23.596711]
 *     ]]
 *   }
 * };
 * var poly2 = {
 *   "type": "Feature",
 *   "properties": {
 *     "fill": "#00f"
 *   },
 *   "geometry": {
 *     "type": "Polygon",
 *     "coordinates": [[
 *       [-46.650009, -23.631314],
 *       [-46.650009, -23.5237],
 *       [-46.509246, -23.5237],
 *       [-46.509246, -23.631314],
 *       [-46.650009, -23.631314]
 *     ]]
 *   }
 * };
 *
 * var deintersected = turf.deintersect([poly1, poly2]);
 *
 * var polygonsBefore = {
 *   "type": "FeatureCollection",
 *   "features": [poly1, poly2]
 * };
 *
 * var polygonsAfter = {
 *   "type": "FeatureCollection",
 *   "features": deintersected
 * };
 *
 * //=polygonsBefore
 *
 * //=polygonsAfter
 */
function differentiate(newPolygons, polygon) {
  const diff = newPolygons.reduce(difference, polygon);
  if (diff !== null) {
    newPolygons.push(diff);
  }
  return newPolygons;
}

function deintersect(polygons) {
  return polygons.slice(1).reduce(differentiate, [polygons[0]]);
}

module.exports = deintersect;
